File: mvn/src/main/java/algorithm/QuickSort.java
package algorithm;

import java.util.Arrays;
import java.util.List;

public class QuickSort {

    static <E extends Comparable> void sort(E[] nums, int start, int end) {
        if(end - start <= 1)
            return;
        int i = start;
        int j = end;
        E flag = nums[start];
        while (i < j){

            while (i < j && nums[i].compareTo(flag) <= 0){
                i++;
            }
            if(i < j){
                nums[j] = nums[i];
                j--;
            }
            while (i < j && nums[j].compareTo(flag) > 0){
                j--;
            }
            if(i < j){
                nums[i] = nums[j];
                i++;
            }
        }
        nums[i] = flag;
        sort(nums, start, i - 1);
        sort(nums, i+1, end);
    }

    static <E> void swap(E a, E b){
        E tmp = a;
        a = b;
        b = tmp;
    }

    public static void main(String[] args) {
        Integer[] s = new Integer[]{5,3,8,1,9,4};
        for (int i = 0; i < s.length; i++) {
            System.out.print(s[i].toString() + " ");
        }
        System.out.println();
        QuickSort.sort(s, 0, s.length - 1);
        for (int i = 0; i < s.length; i++) {
            System.out.print(s[i].toString() + " ");
        }
    }
}

File: algorithm/quicksort.py
def quicksort(nums, r, l):

    if r >= l:
        return

    i = r
    j = l
    flag = nums[r]

    while i < j:
        while nums[j] >= flag and i < j:
            j -= 1
        if i < j:
            nums[i] = nums[j]
            i += 1

        while nums[i] < flag and i < j:
            i += 1
        if i < j:
            nums[j] = nums[i]
            j -= 1
    # print(flag, i)
    nums[i] = flag
    quicksort(nums, r, i-1)
    quicksort(nums, i+1, l)
    # print("out", nums)


if __name__ == '__main__':
    test = [13,8,7,9,15,14,12,15]
    quicksort(test, 0, len(test) -1)
    print(test)




